text/0000-from-iterator-tuple.md

```diff
@@ -0,0 +1,125 @@
+- Feature Name: from-iterator-tuple
+- Start Date: 2017-08-26
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Simplify conversion of `Vec<(_, _)>` and `HashMap<_, _>` to `(Vec<_>, Vec<_>)`.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, there is no way to collect iterator over tuples into two independent collections. Having such `FromIterator` would greatly improve programming ergonomics.
+
+```rust
+struct TupleShunt<'a, I, V> {
+	iter: I,
+	leftovers: &'a mut Vec<V>,
+}
+
+impl<'a, K, V, I> Iterator for TupleShunt<'a, I, V>
+	where
+		I: Iterator<Item = (K, V)>
+{
+	type Item = K;
+
+	fn next(&mut self) -> Option<Self::Item> {
+		match self.iter.next() {
+			Some((k, v)) => {
+				leftovers.push(v);
+				Some(k)
+			},
+			None => None
+		}
+	}
+}
+
+impl<K, V, F, S> FromIterator<(K, V)> for (F, S) 
+	where
+		F: FromIterator<K>,
+		S: FromIterator<V>
+{
+	fn from_iter<T: IntoIterator<(K, V)>>(iter: T) -> (F, S) {
+		let mut leftovers = Vec::new();
+		let mut shunt = TupleShunt {
+			iter: iter.into_iter(),
+			&mut leftovers,
+		};
+		let f = F::from_iter(shunt);
+		let s = S::from_iter(leftovers);
+		(f, s)
+	}
+}
+```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+No new concepts are introduced in this rfc. It's just ergonimics.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+```rust
+fn fetch(x: u32) -> (Foo, Bar);
+
+fn foo(f: Foo) -> ProcessedFoo;
+
+fn bar(b: Bar) -> ProcessedBar;
+
+fn finished_processing_foo(f: Vec<ProcessedFoo>);
+
+fn finished_processing_bar(b: Vec<ProcessedBar>);
+
+fn old_usage() {
+	let foobar: Vec<_> = (0..10)
+		.into_iter()
+		.map(fetch)
+		.map(|(f, b)| (foo(f), bar(b)))
```

Comment:
`FromIterator` is used for collections so far. I'm not sure how I feel about changing it to also work for tuples of collections. I'd rather generalize `unzip` in the future over all tuples if we ever get variadic generics.